src/rust/engine/process_execution/src/remote_cache_tests.rs

```diff
@@ -53,6 +60,7 @@ impl CommandRunnerTrait for MockLocalCommandRunner {
     _req: MultiPlatformProcess,
     _context: Context,
   ) -> Result<FallibleProcessResultWithPlatform, String> {
+    delay_for(self.delay).await;
```

Comment:
Using a delay to control execution of this function will be flaky. Please use a oneshot channel or a [`tokio::sync::Semaphore`](https://docs.rs/tokio/1.0.2/tokio/sync/struct.Semaphore.html) to control the function's execution

Specifically, the `MockLocalCommandRunner` should take an `Arc<Semaphore>` and then just call `.acquire` to obtain a "permit" from the semaphore. The test code should call `add_permits` either before or after the call to `run_process` to setup whether the `.acquire` call blocks or succeeds immediately.

Reply:
Hm I tried doing this just now, but I have no idea how to do this via semaphore, given that `cache_runner.run` possibly uses both the local runner and the StubActionCache. We want to ensure one of those happens before the other.

I was thinking of using the same Semaphore between `MockLocalCommandRunner` and `StubActionCache`, with only one permit, so that we can ensure only one happens at a time. But how do we influence which gains the lock first? I don't think we want to gain/drop the lock in the test code itself, i.e. before or after:

```
    let remote_result = cache_runner
      .run(process.clone().into(), Context::default())
      .await
      .unwrap();
```

All the interesting things happen inside that `cache_runner.run()`.